\documentclass[a4paper]{article}

\usepackage[english]{babel}
\usepackage[utf8]{inputenc}
\usepackage{amsmath}
\usepackage{graphicx}
\usepackage[colorinlistoftodos]{todonotes}
\usepackage{hyperref}
\usepackage{xcolor}
\usepackage{alltt}



\title{DGE pipeline generator}

\author{Patrick Blumenkamp}


\begin{document}
\maketitle

\section{Groups file}
\subsection{Example}
\begin{alltt}
name    forward_reads   reverse_reads   condition
WT01    ../data/WT_01_R1.fastq     ../data/WT_01_R2.fastq     WT
WT02    ../data/WT_02_R1.fastq     ../data/WT_02_R2.fastq     WT
WT03    ../data/WT_02_R1.fastq     ../data/WT_02_R2.fastq     WT
Mut01    ../data/Mut_01_R1.fastq     ../data/Mut_01_R2.fastq     Mut
Mut02    ../data/Mut_02_R1.fastq     ../data/Mut_02_R2.fastq     Mut
Mut03    ../data/Mut_02_R1.fastq     ../data/Mut_02_R2.fastq     Mut
	
\end{alltt}
\section{Settings file}
\subsection{Example}
\begin{alltt}
pipeline:
paired_end: true		\textcolor{blue}{//Must be set to false if single-end data}

preprocessing:
module: ""		\textcolor{blue}{//no preprocessing}

premapping:
modules: ["multiqc"]		\textcolor{blue}{//One premapping module but can be more than one}

mapping:
module: "bowtie2"

bowtie2:	\textcolor{blue}{//Additional settings for the bowtie2 module}
genome_fasta: "reference/listeria_monocytogenes_egd_e.fasta"
alignment_type: "local"

analyses:
modules: ["readxplorer", "count_table", "deseq2"]

count_table:
gff_feature_type: "gene"
gff_feature_name: "ID"
gff_path: "reference/listeria_monocytogenes_egd_e.gff3"

readxplorer:
reference_genome: "reference/listeria_monocytogenes_egd_e.gff3"
readxplorer_cli_path: "../tools/readxplorer/bin/readxplorer-cli"

deseq2:
gff_feature_type: "gene"
gff_feature_name: "ID"
gff_path: "reference/listeria_monocytogenes_egd_e.gff3"
additional_featcounts_options: "-B"
attribute_columns: "'name' 'product'"

\end{alltt}
\section{Modules}
\subsection{Preprocessing}
\subsubsection{gunzip}
'gunzip' is a preprocessing module for decompressing GNU zip archives (.gz).
\paragraph{Output}
The 'preprocessing' directory contains all decompressed fastq input files.
\paragraph{Settings parameter} None.
\paragraph{Group file columns} None.

\subsection{Premapping}
\subsubsection{fastqc}
FastQC is a quality control tool for high throughput sequence data (\url{https://www.bioinformatics.babraham.ac.uk/projects/fastqc/}).
\paragraph{Output}
The module creates a zip archive and a html file for each input file. The results can be viewed with any common web browser via opening the html file.
\paragraph{Settings parameter} None.
\paragraph{Group file columns} None.

\subsubsection{multiqc}
MultiQC collects all FastQC results to visualize them on one website. The 'multiqc' module uses its own FastQC version, so you do not need to include the 'fastqc' module.
\paragraph{Output}
All FastQC results are inside a \emph{fastqc} directory. MultiQC generates one \emph{multiqc\_report.html} file and a \emph{multiqc\_data} directory. To see the MultiQC report, you just need to open the \emph{multiqc\_report.html} file with any common web browser. Additional statistics can be found in the \emph{multiqc\_data} directory.
\paragraph{Settings parameter} None.
\paragraph{Group file columns} None.

\subsection{Mapping}
\subsubsection{bowtie2}
\url{http://bowtie-bio.sourceforge.net/bowtie2/index.shtml}
\paragraph{Output} 
\subparagraph{\{sample name\}.bam(.bai) and \{sample name\}\_unmapped.bam}
For each input file two bam and one bai file will be generated. The alignments are in \emph{\{sample name\}.bam} and all unmapped reads in \emph{\{sample name\}\_unmapped.bam}. The bam file with the aligned reads will automatically get indexed in \emph{\{sample name\}.bam.bai}.
\subparagraph{logs}
\emph{logs} contains the command-line output of each bowtie2 run. 
\subparagraph{stats}
\emph{stats} summarizes the statistics of all bowtie2 runs in one xlsx file.
\paragraph{Settings parameter} 
\subparagraph{genome\_fasta}
File path to the reference genome. Either as an absolute path or relative to the settings file.
\subparagraph{alignment\_type} Local (\emph{local}) or end-to-end (\emph{end-to-end}) alignment.
\subparagraph{additional\_bowtie2\_options (optional)} Additional options to use in the bowtie2 command written as one string, e.g. '--met-stderr --omit-sec-seq'.
\paragraph{Group file columns} None.


\subsection{Analysis}
\subsubsection{deseq2}
\url{https://bioconductor.org/packages/release/bioc/html/DESeq2.html}
\paragraph{Output} 
%\subparagraph{correlation\_heatmap.pdf}
%A heat map showing similarities between all samples. 
\subparagraph{counts.txt and counts.txt.summary}
With featureCounts generated count table of all samples and a summary of the featureCounts run.
%\subparagraph{counts\_assignment.pdf} 
%Visual representation of the featureCounts statistics.
\subparagraph{counts\_normalized.txt}
DESeq2 normalized counts.
\subparagraph{deseq2.RData}
R save state. (Detailed description follows.)
\subparagraph{deseq2\_comparisons}
Each possible comparison between two conditions generates one csv file. Each file contains the DESeq2 results of this comparison.
\subparagraph{logs}
The command-line output of featureCounts and R.
\subparagraph{summary}
For each condition one summary file will be generated. This file compares one condition against all others.
\paragraph{Settings parameter}
\subparagraph{gff\_feature\_type}
Used feature type, e.g. gene, CDS or exon.
\subparagraph{gff\_feature\_name}
Descriptor for gene name, e.g. ID or gene\_id.
\subparagraph{gff\_path}
File path to gff file. Either as an absolute path or relative to the settings file.
\subparagraph{additional\_featcounts\_options}
Additional options to use in featureCounts command, e.g. '--verbose -L'. 
\subparagraph{attribute\_columns}
Listed gff attributes in the beginning of the xlsx summary, e.g. '"name" "product"'
\paragraph{Group file columns}
\subparagraph{condition}
Condition of this sample

\subsubsection{count\_table}
\paragraph{Output} 
\subparagraph{correlation\_heatmap.pdf}
\subparagraph{counts.txt and counts.txt.summary}
With featureCounts generated count table of all samples and a summary of the featureCounts run.
\subparagraph{logs}
\subparagraph{summary}
\paragraph{Settings parameter}
\subparagraph{gff\_feature\_type}
Used feature type, e.g. gene, CDS or exon.
\subparagraph{gff\_feature\_name}
Descriptor for gene name, e.g. ID or gene\_id.
\subparagraph{gff\_path}
File path to gff file. Either as an absolute path or relative to the settings file.
\subparagraph{additional\_options}
Additional options to use in featureCounts command, e.g. '--verbose -L'. 
\paragraph{Group file columns} None.
\subsubsection{readxplorer}
A module for generating a ReadXplorer database. This database can be used to visualize all results in ReadXplorer immediately.
\paragraph{Output} 
\subparagraph{readxplorer.h2.db} ReadXplorer database.
\subparagraph{\{sample name\}\_extended.bam} Used bam files for ReadXplorer database.
\subparagraph{logs} Log files of ReadXplorer import.
\paragraph{Settings parameter}
\subparagraph{readxplorer\_cli\_path}
Path to ReadXplorer-CLI execution file.
\subparagraph{reference\_genome}
Path to reference genome annotation. \textbf{Due to a ReadXplorer bug, this path must be relative.}
\paragraph{Group file columns} None.
\end{document}